// @ngrx/router-store/esm5/src/router_selectors.js
import { createSelector } from '@ngrx/store';
export function getSelectors(selectState) {
    var selectRouterState = createSelector(selectState, function (router) { return router && router.state; });
    var selectCurrentRoute = createSelector(selectRouterState, function (routerState) {
        if (!routerState) {
            return undefined;
        }
        var route = routerState.root;
        while (route.firstChild) {
            route = route.firstChild;
        }
        return route;
    });
    var selectQueryParams = createSelector(selectCurrentRoute, function (route) { return route && route.queryParams; });
    var selectRouteParams = createSelector(selectCurrentRoute, function (route) { return route && route.params; });
    var selectRouteData = createSelector(selectCurrentRoute, function (route) { return route && route.data; });
    var selectUrl = createSelector(selectRouterState, function (routerState) { return routerState && routerState.url; });
    return {
        selectCurrentRoute: selectCurrentRoute,
        selectQueryParams: selectQueryParams,
        selectRouteParams: selectRouteParams,
        selectRouteData: selectRouteData,
        selectUrl: selectUrl,
    };
}
